contrib/docs/tutorials/authenticate-api-requests.md

```diff
@@ -91,10 +78,99 @@ export const createAuthMiddleware = async (
 };
 ```
 
+Create `packages/backend/src/identity.ts`:
+
+```typescript
+import {
+  PluginEndpointDiscovery,
+  TokenManager,
+} from '@backstage/backend-common';
+import {
+  BackstageIdentityResponse,
+  DefaultIdentityClient,
+  getBearerTokenFromAuthorizationHeader,
+  IdentityApi,
+  IdentityApiGetIdentityRequest,
+} from '@backstage/plugin-auth-node';
+
+export class IdentityClientWithServer implements IdentityApi {
+  private inner: IdentityApi;
+  private tokenManager: TokenManager;
+  private constructor({
+    discovery,
+    tokenManager,
+  }: {
+    discovery: PluginEndpointDiscovery;
+    tokenManager: TokenManager;
+  }) {
+    this.inner = DefaultIdentityClient.create({
+      discovery,
+    });
+    this.tokenManager = tokenManager;
+  }
+
+  static create({
+    discovery,
+    tokenManager,
+  }: {
+    discovery: PluginEndpointDiscovery;
+    tokenManager: TokenManager;
+  }) {
+    return new IdentityClientWithServer({ discovery, tokenManager });
+  }
+
+  async getIdentity({
+    request,
+  }: IdentityApiGetIdentityRequest): Promise<
+    BackstageIdentityResponse | undefined
+  > {
+    let identityError: any;
+    try {
+      return await this.inner.getIdentity({ request });
+    } catch (e) {
+      identityError = e;
+    }
+    const token = getBearerTokenFromAuthorizationHeader(
+      request.headers.authorization,
+    );
+    if (!token) {
+      return undefined;
+    }
+    try {
+      await this.tokenManager.authenticate(token);
```

Comment:
I think it'd be better to move this into the middleware instead. The `IdentityApi` is not supposed to be authenticating server identities right now.

Reply:
This comes back to my problem of using server-server tokens for #13991 .

The scaffolder dry-run API calls `getIdentity` to get the `token`, 

https://github.com/backstage/backstage/blob/61f5da108c39306730282e271a11efc0b99034c4/plugins/scaffolder-backend/src/service/router.ts#L500-L504

And the `DefaultIdentityClient` validates the token only against the well-known public keys of the auth API, so that fails.

One way to fix that is to add the server token to the identity api, which really just verifies anyway. Or we'd fix the scaffolder to not fail if can't get the token.

I guess this mostly depends on what the intended use of the IdentityApi is.

Reply:
Hmm, I was sure I had a comment over in #13991 saying that we should avoid using server tokens, or was that somewhere else? TL;DR was that we really shouldn't be using server token to make calls from a CLI tool. Server tokens provide very broad access and should not be thrown around lightly.

We do want to expand the service auth system to support use-cases like that though, but that's not something we have yet, and if that ends up happening we'll end up handling this use-case while at it somehow.

Reply:
I agree that we shouldn't throw server tokens around lightly. The intent for supporting that was to be able to run the generation in CI. Now, that would likely require a custom runner w/ elevated access at least in our case. We don't even have routes open from GH to our instance even. And then all documented concerns around that totally apply.

For the identity part, I could do w/out that, but then I'd need a tweak to the scaffolder router, to ignore failures to get the identity and fall back to a null token.

```diff
diff --git a/plugins/scaffolder-backend/src/service/router.ts b/plugins/scaffolder-backend/src/service/router.ts
index eff7e95c5..5d46c98ac 100644
--- a/plugins/scaffolder-backend/src/service/router.ts
+++ b/plugins/scaffolder-backend/src/service/router.ts
@@ -498,10 +498,10 @@ export async function createRouter(
       }
 
       const token = (
-        await identity.getIdentity({
+        await (identity.getIdentity({
           request: req,
         })
-      )?.token;
+      ).catch(() => null))?.token;
 
       for (const parameters of [template.spec.parameters ?? []].flat()) {
         const result = validate(body.values, parameters);
```

Reply:
Sure - I think this makes sense right? Sorry I've kind of missed some of this identity work here, so struggling for context to be honest. But what @rugvip mentions make sense! Do you want to pick that up in here @pike?

Reply:
I guess there are two ways here, one is to build back to the scaffolder-specific identity implementation. The other is to use the given one, and ignore if it throws errors. That's what I suggested in my patch snippet above.

I guess the difference comes to play if someone actually uses a different scheme for JWT keys, I guess?

Also, the `catch` might end up being less code in the long run?